bobmenu/dialog-0.9a/dialog.1
.TH DIALOG 1 "" "2001/05/17 18:46:10"
.SH NAME
dialog \- display dialog boxes from shell scripts
.SH SYNOPSIS
.B dialog --clear
.br
.BI "dialog --create-rc " file
.br
.B dialog --print-maxsize
.br
.B dialog
.I common-options
.I box-options
.SH DESCRIPTION
.B Dialog
is a program that will let you to present a variety of questions or
display messages using dialog boxes from a shell script.
These types of dialog boxes are implemented
(though not all are necessarily compiled into \fBdialog\fR):
.RS
.LP
.BR calendar ", "
.BR checklist ", "
.BR fselect ", "
.BR gauge ", "
.BR infobox ", "
.BR inputbox ", "
.BR menu ", "
.BR msgbox " (message), "
.BR password ", "
.BR radiolist ", "
.BR tailbox ", "
.BR tailboxbg ", "
.BR textbox ", "
.BR timebox ", and "
.BR yesno " (yes/no)."
.RE
.PP
You can put more than one dialog box into a script:
.TP 5
-
Use the \fB\-\-and-widget\fP token to force Dialog to proceed to the next
dialog unless you have pressed ESC to cancel, or
.TP 5
-
Simply add the tokens for the next dialog box, making a chain.
Dialog stops chaining when the return code from a dialog is nonzero,
e.g., Cancel or No.
.
.\" ************************************************************************
.SH OPTIONS
.TP
.B Common Options
.
.IP "\fB\-\-aspect \fIratio"
This gives you some control over the box dimensions when using auto
sizing (specifying 0 for height and width).  It represents
width / height.  The default is 9, which means 9 characters wide to
every 1 line high.
.
.IP "\fB\-\-backtitle \fIbacktitle"
Specifies a
.I backtitle
string to be displayed on the backdrop, at the top of the screen.
.
.IP "\fB\-\-beep"
Sound the audible alarm
each time the screen is refreshed.
.
.IP "\fB\-\-beep-after"
Beep if input is interrupted, e.g., by a control/C.
.
.IP "\fB\-\-begin \fIy x"
Specify the position of the upper left corner of a dialog box on the screen.
.
.IP "\fB\-\-clear"
The screen will be cleared to the
.B screen attribute
on exit.
This may be used alone, without other options.
.
.IP "\fB\-\-cr-wrap"
Interpret embedded newlines in the dialog text as a newline on the screen.
Otherwise, \fBdialog\fR will only wrap lines where needed to fit inside the text box.
Even though you can control line breaks with this,
\fBdialog\fR will still wrap any lines that are too long for the width of the box.
Without cr-wrap, the layout of your text may be formatted to look nice
in the source code of your script without affecting the way it will
look in the dialog.
.
.IP "\fB\-\-create-rc \fIfile"
When
.B dialog
supports run-time configuration,
this can be used to dump a sample configuration file to the file specified
by
.IR file "."
.
.IP "\fB\-\-help"
Prints the help message to standard error.
The help message is printed if no options are given.
.
.IP "\fB\-\-defaultno"
Make the default value of the
.B yes/no
box a
.BR No .
.
.IP "\fB\-\-default-item \fIstring"
Set the default item in a menu box.
Normally the first item in the box is the default.
.
.IP "\fB\-\-item-help"
Interpret the tags data for checklist, radiolist and menuboxes 
adding a column which is displayed in the bottom line of the
screen, for the currently selected item.
.
.IP "\fB\-\-no-kill"
Tells
.B dialog
to put the
.B tailboxbg
box in the background,
printing its process id to standard error.
SIGHUP is disabled for the background process.
.
.IP "\fB\-\-no-shadow"
Suppress shadows that would be drawn to the right and bottom of each dialog box.
.
.IP "\fB\-\-no-cancel"
.IP "\fB\-\-nocancel"
Suppress the "Cancel" button in checklist, inputbox and menubox modes.
A script can still test if the user pressed the ESC key to cancel to quit.
.
.IP "\fB\-\-print-maxsize"
Print the maximum size of dialog boxes, i.e., the screen size,
to the standard error.
This may be used alone, without other options.
.
.IP "\fB\-\-print-size"
Prints the size of each dialog box to standard error.
.
.IP "\fB\-\-print-version"
Prints \fBdialog\fR's version to standard error.
This may be used alone, without other options.
.
.IP "\fB\-\-separate-output"
For checklist widgets, output result one line at a time, with no
quoting.  This facilitates parsing by another program.
.
.IP "\fB\-\-separate-widget \fIstring"
Specify a string that will separate the output on standard error from
each widget.
This is used to simplify parsing the result of a dialog with several widgets.
.
.IP "\fB\-\-shadow"
Draw a shadow to the right and bottom of each dialog box.
.
.IP "\fB\-\-size-err"
Check the resulting size of a dialog box before trying to use it,
printing the resulting size if it is larger than the screen.
(This option is obsolete, since all new-window calls are checked).
.
.IP "\fB\-\-sleep \fIsecs"
Sleep (delay) for the given number of seconds after processing a dialog box.
.
.IP "\fB\-\-stderr"
Direct output to the standard error.
This is the default, since curses normally writes screen updates to
the standard output.
.
.IP "\fB\-\-stdout"
Direct output to the standard output.
.
.IP "\fB\-\-tab-correct"
Convert each tab character to one or more spaces.
Otherwise, tabs are rendered according to the curses library's interpretation.
.
.IP "\fB\-\-tab-len \fIn"
Specify the number of spaces that a tab character occupies if the
.B "\fB\-\-tab-correct"
option is given.
The default is 8.
.
.IP "\fB\-\-title \fItitle"
Specifies a
.I title
string to be displayed at the top of the dialog box.
.
.IP "\fB\-\-version"
Same as "\fB\-\-print-version\fP".
.
.\" ************************************************************************
.TP
.B Box Options
.RS
All dialog boxes have at least three parameters:
.TP 5
.I text
the caption or contents of the box.
.TP 5
.I height
the height of the dialog box.
.TP 5
.I width
the width of the dialog box.
.PP
Other parameters depend on the box type.
.RE
.IP "\fB\-\-calendar \fItext height width day month year"
A
.B calendar
box displays month, day and year in separately adjustable windows.
You can increment or decrement any of those using the
left-, up-, right- and down-arrows.
Use tab or backtab to move between windows.
If the year is given as zero, the current date is used as an initial value.
On exit, the date is printed in the form day/month/year.
.
.IP "\fB\-\-checklist \fItext height width list-height \fR[ \fItag item status \fR] \fI..."
A
.B checklist
box is similar to a
.B menu
box; there are
multiple entries presented in the form of a menu.
Instead of choosing
one entry among the entries, each entry can be turned on or off by the user.
The initial on/off state of each entry is specified by
.IR status "."
On exit, a list of the
.I tag
strings of those entries that are turned on will be printed on
.IR stderr "."
.
.IP "\fB\-\-fselect \fI text height width\fR"
The file-selection dialog displays a text-entry window in which you can type
a filename (or directory), and above that two windows with directory
names and filenames.
.sp
Here
.B text
can be a filepath in which case the file and directory windows
will display the contents of the path and the text-entry window will contain
the preselected filename.
.sp
Use tab or arrow keys to move between the windows.
Within the directory or filename windows, use the up/down arrow keys
to scroll the current selection.
Use the space-bar to copy the current selection into the text-entry
window.
.sp
Typing any printable characters switches focus to the text-entry window,
entering that character as well as scrolling the directory and filename
windows to the closest match.
.sp
Use a carriage return or the "OK" button to accept the current value
in the text-entry window, or the "Cancel" button to cancel.

.IP "\fB\-\-gauge \fItext height width [percent]\fR"
A
.B gauge
box displays a meter along the bottom of the box.
The meter indicates the percentage.  New percentages are read from
standard input, one integer per line.  The meter is updated
to reflect each new percentage.  If stdin is XXX, then subsequent
lines up to another XXX are used for a new prompt.
The gauge exits when EOF is reached on stdin.
.sp
The \fIpercent\fR value denotes the initial percentage shown in the meter.
If not specified, it is zero.
.
.IP "\fB\-\-infobox \fItext height width"
An
.B info
box is basically a
.B message
box.
However, in this case,
.B dialog
will exit immediately after displaying the message to the user. The
screen is not cleared when
.B dialog
exits, so that the message will remain on the screen until the calling
shell script clears it later. This is useful when you want to inform
the user that some operations are carrying on that may require some
time to finish.
.
.IP "\fB\-\-inputbox \fItext height width [init]"
An
.B input
box is useful when you want to ask questions that
require the user to input a string as the answer. If init is supplied
it is used to initialize the input string.
When entering the
string, the
.I BACKSPACE
key can be used to correct typing errors. If the input string is longer than
can be fitted in the dialog box, the input field will be scrolled. On exit,
the input string will be printed on
.IR stderr "."
.
.IP "\fB\-\-menu \fItext height width menu-height \fR[ \fItag item \fR] \fI..."
As its name suggests, a
.B menu
box is a dialog box that can be used to present a list of choices in
the form of a menu for the user to choose.
Choices are displayed in the order given.
Each menu entry consists of a
.I tag
string and an
.I item
string. The
.I tag
gives the entry a name to distinguish it from the other entries in the
menu. The
.I item
is a short description of the option that the entry represents. The
user can move between the menu entries by pressing the
.I UP/DOWN
keys, the first letter of the
.I tag
as a hot-key, or the number keys
.IR 1-9 ". There are"
.I menu-height
entries displayed in the menu at one time, but the menu will be
scrolled if there are more entries than that. When
.B dialog
exits, the
.I tag
of the chosen menu entry will be printed on
.IR stderr "."
.
.IP "\fB\-\-msgbox \fItext height width"
A
.B message

bobmenu/dialog-0.9a/dialog.c
/*
 *  cdialog - Display simple dialog boxes from shell scripts
 *
 *  AUTHOR: Savio Lam (lam836@cs.cuhk.hk)
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "dialog.h"
#include <string.h>
#ifdef HAVE_SETLOCALE
#include <locale.h>
#endif
#define JUMPARGS const char *t, char *av[], int *offset_add
typedef int (jumperFn) (JUMPARGS);

typedef enum {
    o_unknown = 0
    ,o_and_widget
    ,o_aspect
    ,o_backtitle
    ,o_beep
    ,o_beep_after
    ,o_begin
    ,o_calendar
    ,o_checklist
    ,o_clear
    ,o_cr_wrap
    ,o_create_rc
    ,o_default_item
    ,o_defaultno
    ,o_fselect
    ,o_fullbutton
    ,o_gauge
    ,o_help
    ,o_infobox
    ,o_inputbox
    ,o_item_help
    ,o_menu
    ,o_msgbox
    ,o_no_kill
    ,o_no_shadow
    ,o_nocancel
    ,o_noitem
    ,o_passwordbox
    ,o_print_maxsize
    ,o_print_size
    ,o_print_version
    ,o_radiolist
    ,o_separate_output
    ,o_separate_widget
    ,o_shadow
    ,o_size_err
    ,o_sleep
    ,o_stderr
    ,o_stdout
    ,o_tab_correct
    ,o_tab_len
    ,o_tailbox
    ,o_tailboxbg
    ,o_textbox
    ,o_timebox
    ,o_title
    ,o_yesno
} eOptions;

typedef struct {
    const char *name;
    eOptions code;
    int pass;			/* 1,2,4 or combination */
    const char *help;		/* NULL to suppress, non-empty to display params */
} Options;

typedef struct {
    eOptions code;
    int argmin, argmax;
    jumperFn *jumper;
} Mode;

static const char *program = "dialog";
static const char *and_widget = "--and-widget";
/* *INDENT-OFF* */
static Options options[] = {
    { "and-widget",	o_and_widget,		4, NULL },
    { "aspect",		o_aspect,		1, "<ratio>" },
    { "backtitle",	o_backtitle,		1, "<backtitle>" },
    { "beep",		o_beep,			1, "" },
    { "beep-after",	o_beep_after,		1, "" },
    { "begin",		o_begin,		1, "<y> <x>" },
    { "calendar",	o_calendar,		2, "<text> <height> <width> <day> <month> <year>" },
    { "checklist",	o_checklist,		2, "<text> <height> <width> <list height> <tag1> <item1> <status1>..." },
    { "clear",		o_clear,		1, "" },
    { "cr-wrap",	o_cr_wrap,		1, "" },
    { "create-rc",	o_create_rc,		1, "" },
    { "default-item",	o_default_item,		1, "<str>" },
    { "defaultno",	o_defaultno,		1, "" },
    { "fb",		o_fullbutton,		1, NULL },
    { "fselect",	o_fselect,		2, "<text> <directory> <height> <width>" },
    { "fullbutton",	o_fullbutton,		1, NULL },
    { "gauge",		o_gauge,		2, "<text> <height> <width> [<percent>]" },
    { "guage",		o_gauge,		2, NULL },
    { "help",		o_help,			4, "" },
    { "infobox",	o_infobox,		2, "<text> <height> <width>" },
    { "inputbox",	o_inputbox,		2, "<text> <height> <width> [<init>]" },
    { "item-help",	o_item_help,		1, "" },
    { "menu",		o_menu,			2, "<text> <height> <width> <menu height> <tag1> <item1>..." },
    { "msgbox",		o_msgbox,		2, "<text> <height> <width>" },
    { "no-cancel",	o_nocancel,		1, "" },
    { "no-kill",	o_no_kill,		1, "" },
    { "no-shadow",	o_no_shadow,		1, "" },
    { "nocancel",	o_nocancel,		1, NULL }, /* see --no-cancel */
    { "noitem",		o_noitem,		1, NULL },
    { "passwordbox",	o_passwordbox,		2, "<text> <height> <width> [<init>]" },
    { "print-maxsize",	o_print_maxsize,	1, "" },
    { "print-size",	o_print_size,		1, "" },
    { "print-version",	o_print_version,	5, "" },
    { "radiolist",	o_radiolist,		2, "<text> <height> <width> <list height> <tag1> <item1> <status1>..." },
    { "separate-output",o_separate_output,	1, "" },
    { "separate-widget",o_separate_widget,	1, "<str>" },
    { "shadow",		o_shadow,		1, "" },
    { "size-err",	o_size_err,		1, "" },
    { "sleep",		o_sleep,		1, "<secs>" },
    { "stderr",		o_stderr,		1, "" },
    { "stdout",		o_stdout,		1, "" },
    { "tab-correct",	o_tab_correct,		1, "" },
    { "tab-len",	o_tab_len,		1, "<n>" },
    { "tailbox",	o_tailbox,		2, "<file> <height> <width>" },
    { "tailboxbg",	o_tailboxbg,		2, "<file> <height> <width>" },
    { "textbox",	o_textbox,		2, "<file> <height> <width>" },
    { "timebox",	o_timebox,		2, "<text> <height> <width> <hour> <minute> <second>" },
    { "title",		o_title,		1, "<title>" },
    { "version",	o_print_version,	5, "" },
    { "yesno",		o_yesno,		2, "<text> <height> <width>" },
};
/* *INDENT-ON* */

static eOptions
lookupOption(const char *name, int pass)
{
    unsigned n;

    if (name != 0
	&& !strncmp(name, "--", 2)) {
	name += 2;
	for (n = 0; n < sizeof(options) / sizeof(options[0]); n++) {
	    if ((pass & options[n].pass) != 0
		&& !strcmp(name, options[n].name)) {
		return options[n].code;
	    }
	}
    }
    return o_unknown;
}

static void
Usage(char *msg)
{
    exiterr("Error: %s.\nUse --help to list options.\n\n", msg);
}

/*
 * Count arguments, stopping at the end of the argument list, or on any of our
 * "--" tokens.
 */
static int
arg_rest(char *argv[])
{
    int i = 1;			/* argv[0] points to a "--" token */

    while (argv[i] != 0
	   && (strncmp(argv[i], "--", 2)
	       || lookupOption(argv[i], 7) == o_unknown))
	i++;
    return i;
}

/*
 * In MultiWidget this function is needed to count how many tags
 * a widget (menu, checklist, radiolist) has
 */
static int
howmany_tags(char *argv[], int group)
{
    int result = 0;
    int have;
    const char *format = "Expected %d arguments, found only %d";
    char temp[80];

    while (argv[0] != 0) {
	if (!strcmp(argv[0], and_widget))
	    break;
	if ((have = arg_rest(argv)) < group) {
	    sprintf(temp, format, group, have);
	    Usage(temp);
	}

	argv += group;
	result++;
    }
    if (argv[0] != 0
	&& strcmp(argv[0], and_widget) != 0
	&& (have = arg_rest(argv)) < group) {
	sprintf(temp, format, group, have);
	Usage(temp);
    }

    return result;
}

static char *
optional_str(char **av, int n, char *dft)
{
    char *ret = dft;
    if (arg_rest(av) > n)
	ret = av[n];
    return ret;
}

static int
optional_num(char **av, int n, int dft)
{
    int ret = dft;
    if (arg_rest(av) > n)
	ret = atoi(av[n]);
    return ret;
}

/*
 * These are the program jumpers
 */

static int
j_yesno(JUMPARGS)
{
    *offset_add = 4;
    return dialog_yesno(t,
			av[1],
			atoi(av[2]),
			atoi(av[3]), defaultno);
}

static int
j_msgbox(JUMPARGS)
{
    *offset_add = 4;
    return dialog_msgbox(t,
			 av[1],
			 atoi(av[2]),
			 atoi(av[3]), 1);
}

static int
j_infobox(JUMPARGS)
{
    *offset_add = 4;
    return dialog_msgbox(t,
			 av[1],
			 atoi(av[2]),
			 atoi(av[3]), 0);
}

static int
j_textbox(JUMPARGS)
{
    *offset_add = 4;
    return dialog_textbox(t,
			  av[1],
			  atoi(av[2]),
			  atoi(av[3]));
}

/* MAC: from menubox.c */
extern char NUMERIC_INPUT[256];

static int
j_menu(JUMPARGS)
{
    int ret;
    int tags = howmany_tags(av + 5, MENUBOX_TAGS);
    *offset_add = 5 + tags * MENUBOX_TAGS;
    ret = dialog_menu(t,
		      av[1],
		      atoi(av[2]),
		      atoi(av[3]),
		      atoi(av[4]),
		      tags, av + 5);
    if (ret >= 0) {
	fprintf(dialog_vars.output, av[5 + ret * MENUBOX_TAGS]);
	return 0;
    } else if (ret == -2) {
	return 1;		/* CANCEL */
    } else if (ret == -3) {
      /* MAC: return the numeric string  */
      fprintf(dialog_vars.output, NUMERIC_INPUT);
      return 0;
    }
    return ret;			/* ESC pressed, ret == -1 */
}

static int
j_checklist(JUMPARGS)
{
    int tags = howmany_tags(av + 5, CHECKBOX_TAGS);
    *offset_add = 5 + tags * CHECKBOX_TAGS;
    return dialog_checklist(t,
			    av[1],
			    atoi(av[2]),
			    atoi(av[3]),
			    atoi(av[4]),
			    tags, av + 5, FLAG_CHECK, dialog_vars.separate_output);
}

static int
j_radiolist(JUMPARGS)
{
    int tags = howmany_tags(av + 5, CHECKBOX_TAGS);
    *offset_add = 5 + tags * CHECKBOX_TAGS;
    return dialog_checklist(t,
			    av[1],
			    atoi(av[2]),
			    atoi(av[3]),
			    atoi(av[4]),
			    tags, av + 5, FLAG_RADIO, dialog_vars.separate_output);
}

static int
j_inputbox(JUMPARGS)
{
    int ret;

    *offset_add = arg_rest(av);
    ret = dialog_inputbox(t,
			  av[1],
			  atoi(av[2]),
			  atoi(av[3]),
			  optional_str(av, 4, 0), 0);
    if (ret == 0)
	fprintf(dialog_vars.output, "%s", dialog_input_result);
    return ret;
}

static int
j_passwordbox(JUMPARGS)
{
    int ret;

    *offset_add = arg_rest(av);
    ret = dialog_inputbox(t,
			  av[1],
			  atoi(av[2]),
			  atoi(av[3]),
			  optional_str(av, 4, 0), 1);
    if (ret == 0)
	fprintf(dialog_vars.output, "%s", dialog_input_result);
    return ret;
}

#ifdef HAVE_XDIALOG
static int
j_calendar(JUMPARGS)
{
    int ret;

    *offset_add = arg_rest(av);
    ret = dialog_calendar(t,
			  av[1],
			  atoi(av[2]),
			  atoi(av[3]),
			  atoi(av[4]),
			  atoi(av[5]),
			  atoi(av[6]));
    if (ret == 0)
	fprintf(dialog_vars.output, "%s", dialog_input_result);
    return ret;
}

static int
j_fselect(JUMPARGS)
{
    int ret;

    *offset_add = arg_rest(av);
    ret = dialog_fselect(t,
			 av[1],
			 atoi(av[2]),
			 atoi(av[3]));
    if (ret == 0)
	fprintf(dialog_vars.output, "%s", dialog_input_result);
    return ret;
}

static int
j_timebox(JUMPARGS)
{
    int ret;

    *offset_add = arg_rest(av);
    ret = dialog_timebox(t,
			 av[1],
			 atoi(av[2]),
			 atoi(av[3]),
			 optional_num(av, 4, -1),
			 optional_num(av, 5, -1),
			 optional_num(av, 6, -1));
    if (ret == 0)
	fprintf(dialog_vars.output, "%s", dialog_input_result);
    return ret;
}
#endif

#ifdef HAVE_GAUGE
static int
j_gauge(JUMPARGS)
{
    *offset_add = arg_rest(av);
    return dialog_gauge(t,
			av[1],
			atoi(av[2]),
			atoi(av[3]),
			optional_num(av, 4, 0));
}
#endif

#ifdef HAVE_TAILBOX
static int
j_tailbox(JUMPARGS)
{
    *offset_add = 4;
    return dialog_tailbox(t,
			  av[1],
			  atoi(av[2]),
			  atoi(av[3]));
}

static int
j_tailboxbg(JUMPARGS)
{
    if (tailbg_lastpid + 1 >= MAX_TAILBG) {
	char temp[80];
	sprintf(temp, "lastpid value %d is greater than %d",
		(int) tailbg_lastpid, MAX_TAILBG);
	Usage(temp);
    }

    *offset_add = 4;

    /*
     * Try to avoid having stdout from the parent process's initialization of
     * curses mess up the display from a backgrounded process.
     */
    if (dialog_vars.cant_kill) {
	(void) refresh();
	(void) fflush(stdout);
    }

    tailbg_pids[tailbg_lastpid] = fork();
    if (dialog_vars.cant_kill)
	tailbg_nokill_pids[tailbg_nokill_lastpid] =
	    tailbg_pids[tailbg_lastpid];

    /* warning! here are 2 processes */

    if (tailbg_pids[tailbg_lastpid] == 0) {
	if (dialog_vars.cant_kill)
	    fprintf(dialog_vars.output, "%d", (int) getpid());
	is_tailbg = TRUE;
	dialog_tailboxbg(t,
			 av[1],
			 atoi(av[2]),
			 atoi(av[3]),
			 dialog_vars.cant_kill);

	(void) refresh();	/* quitting for SIGHUP and --no-kill */
	end_dialog();
	exit(0);
    }
    if (dialog_vars.cant_kill)
	tailbg_nokill_lastpid++;
    tailbg_lastpid++;
    return 0;
}
#endif
/* *INDENT-OFF* */
static Mode modes[] =
{
    {o_yesno, 4, 4, j_yesno},
    {o_msgbox, 4, 4, j_msgbox},
    {o_infobox, 4, 4, j_infobox},
    {o_textbox, 4, 4, j_textbox},
    {o_menu, 7, 0, j_menu},
    {o_checklist, 8, 0, j_checklist},
    {o_radiolist, 8, 0, j_radiolist},
    {o_inputbox, 4, 5, j_inputbox},
    {o_passwordbox, 4, 5, j_passwordbox},
#ifdef HAVE_XDIALOG
    {o_calendar, 7, 7, j_calendar},
    {o_fselect, 4, 5, j_fselect},
    {o_timebox, 4, 7, j_timebox},
#endif
#ifdef HAVE_GAUGE
    {o_gauge, 4, 5, j_gauge},
#endif
#ifdef HAVE_TAILBOX
    {o_tailbox, 4, 4, j_tailbox},
    {o_tailboxbg, 4, 4, j_tailboxbg},
#endif
};
/* *INDENT-ON* */

static char *
optionString(char **argv, int *num)
{
    int next = *num + 1;
    char *result = argv[next];
    if (result == 0) {
	char temp[80];
	sprintf(temp, "Expected a string-parameter for %.20s", argv[*num]);
	Usage(temp);
    }
    *num = next;
    return result;
}

static int
optionValue(char **argv, int *num)
{
    int next = *num + 1;
    char *src = argv[next];
    char *tmp = 0;
    int result = 0;

    if (src != 0) {
	result = strtol(src, &tmp, 0);
	if (tmp == 0 || *tmp != 0)
	    src = 0;
    }

    if (src == 0) {
	char temp[80];
	sprintf(temp, "Expected a numeric-parameter for %.20s", argv[*num]);
	Usage(temp);
    }
    *num = next;
    return result;
}

/*
 * Print parts of a message
 */
static void
PrintList(const char **list)
{
    const char *leaf = strrchr(program, '/');
    unsigned n = 0;

    if (leaf != 0)
	leaf++;
    else
	leaf = program;

    while (*list != 0) {
	fprintf(dialog_vars.output, *list, n ? leaf : VERSION);
	(void) fputc('\n', dialog_vars.output);
	n = 1;
	list++;
    }
}

static Mode *
lookupMode(eOptions code)
{
    Mode *modePtr = 0;
    unsigned n;
    for (n = 0; n < sizeof(modes) / sizeof(modes[0]); n++) {
	if (modes[n].code == code) {
	    modePtr = &modes[n];
	    break;
	}
    }
    return modePtr;
}

/*
 * Print program help-message
 */
static void
Help(void)
{
    static const char *tbl_1[] =
    {
	"cdialog (ComeOn Dialog!) version %s",
	"",
	"* Display dialog boxes from shell scripts *",
	"",
	"Usage: %s <options> { --and-widget <options> }",
	"where options are \"common\" options, followed by \"box\" options",
	"",
#ifdef HAVE_RC_FILE
	"Special options:",
	"  [--create-rc \"Ifile\"]",
#endif
	0
    }, *tbl_3[] =
    {
	"",
	"Auto-size with height and width = 0. Maximize with height and width = -1.",
	"Global-auto-size if also menu_height/list_height = 0.",
	0
    };
    unsigned j, k;

    PrintList(tbl_1);
    fprintf(dialog_vars.output, "Common options:\n ");
    for (j = k = 0; j < sizeof(options) / sizeof(options[0]); j++) {
	if ((options[j].pass & 1)
	    && options[j].help != 0) {
	    unsigned len = 6 + strlen(options[j].name) + strlen(options[j].help);
	    k += len;
	    if (k > 75) {
		fprintf(dialog_vars.output, "\n ");
		k = len;
	    }
	    fprintf(dialog_vars.output, " [--%s%s%s]", options[j].name,
		    *(options[j].help) ? " " : "", options[j].help);
	}
    }
    fprintf(dialog_vars.output, "\nBox options:\n");
    for (j = 0; j < sizeof(options) / sizeof(options[0]); j++) {
	if ((options[j].pass & 2) != 0
	    && options[j].help != 0
	    && lookupMode(options[j].code))
	    fprintf(dialog_vars.output, "  --%-12s %s\n", options[j].name,
		    options[j].help);
    }
    PrintList(tbl_3);

    exit(EXIT_SUCCESS);
}

int
main(int argc, char *argv[])
{
    FILE *output = stderr;
    char temp[80];
    const char *separate_str = DEFAULT_SEPARATE_STR;
    int esc_pressed = 0;
    int offset = 1;
    int offset_add;
    int retval = 0;
    int done;
    int j;
    eOptions code;
    Mode *modePtr;
#ifndef HAVE_COLOR
    int use_shadow = FALSE;	/* ignore corresponding option */
#endif

#if defined(ENABLE_NLS)
    /* initialize locale support */
    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);
#elif defined(HAVE_SETLOCALE)
    (void) setlocale(LC_ALL, "");
#endif

    program = argv[0];
    dialog_vars.output = output;

    if (argc == 2) {		/* if we don't want clear screen */
	switch (lookupOption(argv[1], 7)) {
	case o_print_maxsize:
	    (void) initscr();
	    fprintf(output, "MaxSize: %d, %d\n", SLINES, SCOLS);
	    end_dialog();
	    break;
	case o_print_version:
	    fprintf(output, "Version: %s\n", VERSION);
	    break;
	default:
	case o_help:
	    Help();
	    break;
	}
	return 0;
    }

    if (argc < 2) {
	Help();
    }

    init_dialog();

#ifdef HAVE_RC_FILE
    if (!strcmp(argv[1], "--create-rc")) {
	if (argc != 3) {
	    sprintf(temp, "Expected a filename for %s", argv[1]);
	    Usage(temp);
	}
	end_dialog();
	create_rc(argv[2]);
	return 0;
    }
#endif
    if ((make_lock_filename(&lock_refresh, "file")) < 0 ||
	(make_lock_filename(&lock_tailbg_refreshed, "tail")) < 0 ||
	(make_lock_filename(&lock_tailbg_exit, "exit")) < 0)
	exiterr("Internal error: can't make lock files.");

    while (offset < argc && !esc_pressed) {
	memset(&dialog_vars, 0, sizeof(dialog_vars));
	dialog_vars.aspect_ratio = DEFAULT_ASPECT_RATIO;
	dialog_vars.tab_len = TAB_LEN;
	dialog_vars.output = output;
	done = FALSE;

	while (offset < argc && !done) {	/* Common options */
	    switch (lookupOption(argv[offset], 1)) {
	    case o_title:
		dialog_vars.title = optionString(argv, &offset);
		break;
	    case o_backtitle:
		dialog_vars.backtitle = optionString(argv, &offset);
		break;
	    case o_separate_widget:
		separate_str = optionString(argv, &offset);
		break;
	    case o_separate_output:
		dialog_vars.separate_output = TRUE;
		break;
	    case o_cr_wrap:
		dialog_vars.cr_wrap = TRUE;
		break;
	    case o_no_kill:
		dialog_vars.cant_kill = TRUE;
		break;
	    case o_nocancel:
		dialog_vars.nocancel = TRUE;
		break;
	    case o_size_err:
		dialog_vars.size_err = TRUE;
		break;
	    case o_beep:
		dialog_vars.beep_signal = TRUE;
		break;
	    case o_beep_after:
		dialog_vars.beep_after_signal = TRUE;
		break;
	    case o_shadow:
		use_shadow = TRUE;
		break;
	    case o_defaultno:
		defaultno = TRUE;
		break;
	    case o_default_item:
		dialog_vars.default_item = optionString(argv, &offset);
		break;
	    case o_item_help:
		dialog_vars.item_help = TRUE;
		break;
	    case o_no_shadow:
		use_shadow = FALSE;
		break;
	    case o_print_size:
		dialog_vars.print_siz = TRUE;
		break;
	    case o_print_maxsize:
		fprintf(output, "MaxSize: %d, %d\n", SLINES, SCOLS);
		break;
	    case o_print_version:
		fprintf(output, "Version: %s\n", VERSION);
		break;
	    case o_tab_correct:
		dialog_vars.tab_correct = TRUE;
		break;
	    case o_sleep:
		dialog_vars.sleep_secs = optionValue(argv, &offset);
		break;
	    case o_stderr:
		dialog_vars.output = output = stderr;
		break;
	    case o_stdout:
		dialog_vars.output = output = stdout;
		break;
	    case o_tab_len:
		dialog_vars.tab_len = optionValue(argv, &offset);
		break;
	    case o_aspect:
		dialog_vars.aspect_ratio = optionValue(argv, &offset);
		break;
	    case o_begin:
		dialog_vars.begin_set = TRUE;
		dialog_vars.begin_y = optionValue(argv, &offset);
		dialog_vars.begin_x = optionValue(argv, &offset);
		break;
	    case o_clear:
		if (argc == 2) {	/* we only want to clear the screen */
		    killall_bg();
		    (void) refresh();
		    end_dialog();
		    return 0;
		}
		dialog_vars.clear_screen = TRUE;
		break;
	    case o_noitem:
	    case o_fullbutton:
		/* ignore */
		break;
	    default:		/* no more common options */
		done = TRUE;
		break;
	    }
	    if (!done)
		offset++;
	}

	for (j = 1; j < argc; j++) {
	    if (strcmp(argv[j - 1], "--backtitle") != 0 &&
		strcmp(argv[j - 1], "--title") != 0)
		dlg_trim_string(argv[j]);
	}

	if (argv[offset] == NULL) {
	    Usage("Expected a box option");
	}

	if (lookupOption(argv[offset], 2) != o_checklist
	    && dialog_vars.separate_output) {
	    sprintf(temp, "Expected --checklist, not %.20s", argv[offset]);
	    Usage(temp);
	}

	if (dialog_vars.aspect_ratio == 0)
	    dialog_vars.aspect_ratio = DEFAULT_ASPECT_RATIO;

	put_backtitle();

	/* use a table to look for the requested mode, to avoid code duplication */

	modePtr = 0;
	if ((code = lookupOption(argv[offset], 2)) != o_unknown)
	    modePtr = lookupMode(code);
	if (modePtr == 0) {
	    sprintf(temp, "Unknown option %.20s", argv[offset]);
	    Usage(temp);
	}

	if (arg_rest(&argv[offset]) < modePtr->argmin) {
	    sprintf(temp, "Expected at least %d tokens for %.20s, have %d",
		    modePtr->argmin, argv[offset], arg_rest(&argv[offset]));
	    Usage(temp);
	}
	if (modePtr->argmax && arg_rest(&argv[offset]) > modePtr->argmax) {
	    sprintf(temp,
		    "Expected no more than %d tokens for %.20s, have %d",
		    modePtr->argmax, argv[offset], arg_rest(&argv[offset]));
	    Usage(temp);
	}

	retval = (*(modePtr->jumper)) (dialog_vars.title, argv + offset, &offset_add);
	offset += offset_add;

	if (retval == -1) {
	    esc_pressed = 1;
	} else {

	    if (dialog_vars.beep_after_signal)
		(void) beep();

	    if (dialog_vars.sleep_secs)
		(void) napms(dialog_vars.sleep_secs * 1000);

	    if (offset < argc) {
		switch (lookupOption(argv[offset], 7)) {
		case o_and_widget:
		    (void) fputs(separate_str, output);
		    offset++;
		    break;
		case o_unknown:
		    sprintf(temp, "Expected --and-widget, not %.20s",
			    argv[offset]);
		    Usage(temp);
		    break;
		default:
		    /* if we got a cancel, etc., stop chaining */
		    if (retval != 0)
			esc_pressed = 2;
		    else
			dialog_vars.clear_screen = TRUE;
		    break;
		}
	    }
	    if (dialog_vars.clear_screen)
		dialog_clear();
	}

    }

    /*
     * If we have background processes running, avoid resetting the terminal
     * characteristics, let them do it when they exit.
     */
    if (quitall_bg())
	_exit(retval);

    (void) refresh();
    end_dialog();
    return retval;
}
/* End of main() */
